bitreverse_exprt: Expression to reverse the order of bits

tautschnig opened the pull request:
Clang has a __builtin_bitreverse (and at some point GCC might as well:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=50481).

Creating this as a draft: the actual implementation in the back-end is yet to be added, as are tests.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
bitreverse_exprt: Expression to reverse the order of bits

Clang has a __builtin_bitreverse (and at some point GCC might as well:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=50481).

chrisr-diffblue commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -3238,6 +3238,27 @@ exprt c_typecheck_baset::do_special_functions(
+    // clang only
+    if(expr.arguments().size() != 1)
+    {
+      error().source_location = f_op.source_location();
+      error() << identifier << " expects one operand" << eom;
+      throw 0;
```
Throw a structured exception?
```suggestion
       std::ostringstream error_message;
      error_message << expr.source_location().as_string() << ": " << identifier
                    << " expects expects one operand";
      throw invalid_source_file_exceptiont{error_message.str()};
```

tautschnig replied:
Done, but I really wish we had #5837 (with its two dependencies) in place to make things even more structured!

tautschnig replied:
And now also includes a test for this one :-)

chrisr-diffblue commented on src/ansi-c/expr2c.cpp:
```diff
@@ -3496,6 +3496,22 @@ std::string expr2ct::convert_conditional_target_group(const exprt &src)
+    {
+      return convert_function(
+        src, "__builtin_bitreverse" + std::to_string(width));
+    }
```
Belt-and-braces, add an `UNREACHABLE` or an assertion to be more descriptive, instead of using `convert_norep`? Or maybe just have a `PRECONDITION` ?

tautschnig replied:
Hmm, there isn't really anything fundamentally wrong in creating a `bitreverse_exprt` with a type that isn't 8/16/32/64 bits wide, we just don't have a good C representation available. So I'd stick with `convert_norep` as that's what the remainder of this code does, if that's ok?

chrisr-diffblue replied:
Fine by me - it was just a sanity check query from me, thanks.

chrisr-diffblue reviewed (commented):
Generally looks great to me. Given the effort that went into removing unstructure exceptions in the error reporting previously, it seems a shame to be adding a `throw 0` back in here - so could I kindly ask if that could be replaced with a structured exception? It'd be an approval from me if that were done :-)

chrisr-diffblue reviewed (approved):
Many thanks for the updates - looks great to me... And given that I poked a raw nerve on the structured exception point - feel free to assign any PRs that need reviews/comments on that front to me. 👍

kroening commented on src/solvers/flattening/boolbv.cpp:
```diff
@@ -155,11 +155,14 @@ bvt boolbvt::convert_bitvector(const exprt &expr)
+    expr.id() == ID_bitxor || expr.id() == ID_bitxnor ||
+    expr.id() == ID_bitnor || expr.id() == ID_bitnand ||
+    expr.id() == ID_bitreverse)
+  {
```
bit-reversing is not a bit-wise operation. Bit-wise operations are functions on individual bits, not bit-vectors. I'd stick this into a separate file.

tautschnig replied:
Done!

kroening commented on src/util/bitvector_expr.cpp:
```diff
@@ -142,3 +142,17 @@ exprt count_trailing_zeros_exprt::lower() const
+  const exprt operand = op();
+  for(std::size_t i = 0; i < int_width; ++i)
+    result_bits.push_back(extractbit_exprt{operand, i});
+
```
This is potentially problematic when the lowered expression is passed to any piece of code that doesn't cope well with the sharing. This is worth at least a warning, and it may possibly make sense to fix this issue by introducing a `let_exprt` for `op()`.

tautschnig replied:
I've made a stab at it - is the second commit that I added what you were looking for?

kroening replied:
yes!

tautschnig committed:
Bitreverse expression: use let expression in lowering

If a back-end isn't sharing aware, repeated occurrence of the
operand-to-reverse could be costly.
